Introduce new commands

bacongobbler opened the pull request:
This restructures the library and introduces several new commands:

- hippo app <add|remove>
- hippo auth <register|login|logout|whoami>
- hippo certificate <add|remove>
- hippo channel <add|remove>
- hippo env <add|remove>
- hippo revision add

These commands are now possible thanks to the API updates provided in Hippo 0.2.

The packages received an overhaul as well:

- all files preceded with `bindle_` were moved to module `bindle`
- the CLI is now hosted in module `cli`
- all commands are defined in module `cli::commands`
- the hippo client has moved to module `hippo`

closes #31

Signed-off-by: Matthew Fisher <matt.fisher@fermyon.com>

bacongobbler commented:
NOTE: only `hippo auth register` is partially implemented at the moment, but the majority of the skeleton framework has been laid out. Now I have to go back and implement the commands.

radu-matei commented on src/cli/commands/mod.rs:
```diff
@@ -0,0 +1,98 @@
+pub(crate) enum Commands {
+    /// Add, update, and remove Applications
+    #[clap(subcommand)]
+    App(app::Commands),
```
nit: would it be more readable if command enums were directly named like `AppCommands` or `AuthCommands`?

bacongobbler replied:
I figured the namespace would be enough of an indicator… Is the convention in Rust to “stutter” and re-repeat the namespace in the type name (`app::AppCommands`)? I thought most language linters discourage that convention

radu-matei replied:
You would `use app::AppCommands` in the top level, and throughout the file you'd use `AppCommands` directly.

bacongobbler replied:
Clippy has a linting rule against types that stutter (`clippy::module_name_repetitions`), but it is disabled by default. Names like `app::AppCommands` is generally discouraged.

radu-matei commented on src/cli/commands/auth.rs:
```diff
@@ -0,0 +1,48 @@
+        username: Option<String>,
+        /// The password
+        #[clap(long)]
+        password: Option<String>,
```
Also, when the validation fails without an exception, the CLI will return a success message, even if the account hasn't been register successfully:

```
❯ hippo auth register --username root --password PassWord! --danger-accept-invalid-certs https://localhost:5309
Registered new account: root
```

Server logs:

```
info: Hippo.Application.Accounts.Commands.CreateAccountCommand[0]
      Hippo Request: CreateAccountCommand   Hippo.Application.Accounts.Commands.CreateAccountCommand
warn: Microsoft.AspNetCore.Identity.UserManager[14]
      User password validation failed: PasswordRequiresDigit.
```

bacongobbler replied:
Yeah this is a server-side issue. Hippo should be returning a proper error response. We can fix that up in the API :)

bacongobbler replied:
The latter case is also a server-side issue. Nice catch.

bacongobbler replied:
> Should there be some local validation here before sending to the server, as well as on the server?

I don't see a reason why we should validate locally... That just means we'll have to update the same validation rules in two different locations, and that could lead to validation drift. IMO I'd rather put this validation all on the server so every client validates objects the same.

bacongobbler replied:
filed https://github.com/deislabs/hippo/issues/398 and https://github.com/deislabs/hippo/issues/399 - thanks!

bacongobbler replied:
New error message:

```
><> hippo auth register --danger-accept-invalid-certs
Enter username: bacongobbler
Enter password: [hidden]
Error: {"type":"https://tools.ietf.org/html/rfc7231#section-6.5.1","title":"One or more validation errors occurred.","status":400,"errors":{"Password":["The length of 'Password' must be at least 8 characters. You entered 3 characters."]}}
```

radu-matei commented on src/cli/commands/app.rs:
```diff
@@ -0,0 +1,17 @@
+use clap::Subcommand;
+
+#[derive(Subcommand)]
+pub(crate) enum Commands {
```
Not blocking this PR — do we also want a way to list the apps?

bacongobbler replied:
Yes, though there’s no support for that in the API at this time. I also have to expose /edit APIs much like what you can do in the web UI

radu-matei commented on src/cli/commands/app.rs:
```diff
@@ -0,0 +1,17 @@
+    /// Remove an application
+    Remove {
+        /// The application ID
+        id: String,
```
Creating an application works great — I can validate that in the UI:

```
❯ hippo app add testapp radu/testapp
Added App testapp (ID = '2f0c046f-ad29-4846-892b-b850e7d0eba7')
IMPORTANT: save this App ID for later - you will need it to update and/or delete the App (for now)
```

Deleting the app doesn't seem to work:

```
❯ hippo app remove 2f0c046f-ad29-4846-892b-b850e7d0eba7
Error: error in response: status code 404 Not Found
```

bacongobbler replied:
Good catch. I think a 404 is the expected response, so maybe we need to capture that instead of casting via `?`

bacongobbler replied:
Hmm... The server should be responding with a 204 no content according to https://github.com/deislabs/hippo/blob/bae252ef73038ec6e473d852c384aef70695e84d/src/Web/Api/AppController.cs#L41

Will look into this further

bacongobbler replied:
This is happening in the UI as well, so I'm lead to believe this is a server-side issue. Once I determine the cause I'll try to fix it before merging this

bacongobbler replied:
identified the issue in https://github.com/deislabs/hippo/pull/401 - misconfigured API routes

radu-matei commented on src/cli/commands/environment_variable.rs:
```diff
@@ -0,0 +1,17 @@
+        /// The channel ID this environment variable will be bound to
+        channel_id: String,
+    },
+    Remove {
```
Creating a channel works, but removing it returns a 404:

```
❯ hippo channel add staging 2f0c046f-ad29-4846-892b-b850e7d0eba7
Added Channel staging (ID = 'e5114347-fba8-4d19-93c0-59a4cd3e54b4')
IMPORTANT: save this Channel ID for later - you will need it to update and/or delete the Channel (for now)
```

```
❯ hippo channel remove e5114347-fba8-4d19-93c0-59a4cd3e54b4
Error: error in response: status code 404 Not Found
```

bacongobbler replied:
https://github.com/deislabs/hippo-cli/pull/99#discussion_r794740312

radu-matei commented on src/cli/commands/channel.rs:
```diff
@@ -0,0 +1,26 @@
+        #[clap(long)]
+        certificate_id: Option<String>,
+    },
+    Remove {
```
Creating a channel works, but removing it returns a 404:

```
❯ hippo channel add staging 2f0c046f-ad29-4846-892b-b850e7d0eba7
Added Channel staging (ID = 'e5114347-fba8-4d19-93c0-59a4cd3e54b4')
IMPORTANT: save this Channel ID for later - you will need it to update and/or delete the Channel (for now)

❯ hippo channel remove e5114347-fba8-4d19-93c0-59a4cd3e54b4
Error: error in response: status code 404 Not Found
```

bacongobbler replied:
https://github.com/deislabs/hippo-cli/pull/99#discussion_r794740312

bacongobbler commented on src/cli/mod.rs:
```diff
@@ -0,0 +1,538 @@
+            danger_accept_invalid_certs: false,
+            hippo_token_info: None,
+            hippo_username: "".to_owned(),
+            hippo_url: "https://localhost:5309".to_owned(),
```
good call. Done

bacongobbler commented on src/cli/commands/channel.rs:
```diff
@@ -0,0 +1,26 @@
+    Add {
+        name: String,
+
+        app_id: String,
```
fixed

bacongobbler commented:
Happy path workflow:

```
$ hippo auth register --danger-accept-invalid-certs
Enter username: bacongobbler
Enter password: [hidden]
Registered bacongobbler
$ hippo auth login --danger-accept-invalid-certs
Enter Hippo username: bacongobbler
Enter Hippo password: [hidden]
Logged in as bacongobbler
$ hippo app add toast-on-demand toast-on-demand
Added App toast-on-demand (ID = 'fbfc8c40-7e39-4ce7-8693-ee7643f564a0')
IMPORTANT: save this App ID for later - you will need it to update and/or delete the App (for now)
$ hippo channel add latest fbfc8c40-7e39-4ce7-8693-ee7643f564a0
Added Channel latest (ID = '29f28347-9f5d-4323-92dc-5a7172afff8a')
IMPORTANT: save this Channel ID for later - you will need it to update and/or delete the Channel (for now)
$ hippo push . -v production
Pushed toast-on-demand/0.3.0
$ curl -H "Host: latest.toast-on-demand.hippofactory.io" https://localhost:5309 -k
<h1>Welcome to the Toast on Demand web site</h1>
<p>Our mission is to provide you with delicious and nutritious toast - where you want it, when you want it, and as browned as you want it.</p>
<h2>TOAST OF THE DAY</h2>
<p>Today's special is RAW TOAST</p>
<p>Served straight or on the rocks.</p>
<p><img src='assets/images/raw_toast.jpg' /></p>
```

radu-matei reviewed (approved):
I think all the comments were addressed.

LGTM!
